File: src/sort/quickSort.cc

#include <stdio.h>

#include <vector>

using std::vector;

class Solution {
public:
  vector<int> quick_sort(vector<int>& nums)
  {
    quick_sort_c(nums, 0, nums.size() - 1);
    return nums;
  }

  void quick_sort_c(vector<int>& nums, int left, int right)
  {
    if(left >= right) return;

    int pivot = partition(nums, left, right);
    quick_sort_c(nums, left, pivot - 1);
    quick_sort_c(nums, pivot + 1, right);
  }

  int partition(vector<int>& nums, int left, int right)
  {
    auto pivot = nums[right];
    auto i = left;
    for (size_t j = left; j < right; j++) {
      if(nums[j] < pivot) {
        std::swap(nums[i], nums[j]);
        i++;
      }
    }

    std::swap(nums[i], nums[right]);
    return i;
  }
};

int main()
{
  vector<int> input{5,1,1,2,0,0};
  Solution solution;
  auto res = solution.quick_sort(input);
  getchar();
}

File: src/sort/go/quick_sort.go
package main

//递推公式
//quick_sort(p...r) = quick_sort(p...q-1) + quick_sort(q+1...r)

// func main() {
// 	nums := []int{6, 4, 2, 5, 3, 1}
// 	fmt.Println(nums)
// 	quick_sort(nums)
// 	fmt.Println(nums)
// }

func quick_sort(nums []int) {
	if len(nums) < 2 {
		return
	}

	left := 0
	right := len(nums) - 1

	quick_sort_c(nums, left, right)
}

func quick_sort_c(nums []int, left, right int) {
	if left >= right {
		return
	}

	mid := partition(nums, left, right)
	quick_sort_c(nums, left, mid-1)
	quick_sort_c(nums, mid+1, right)
}

func partition(nums []int, left, right int) int {
	pivot := nums[right]
	i := left
	j := left

	for i < right && j < right {
		if nums[j] < pivot {
			nums[i], nums[j] = nums[j], nums[i]
			i++
			j++
		} else {
			j++
		}
	}

	nums[i], nums[right] = nums[right], nums[i]
	return i
}
